cli/gtk: add +new-window action

jcollie opened the pull request:
This will (on GTK) use a D-Bus method call to tell a running Ghostty instance to open a new window. If D-Bus activation is configured properly, Ghostty does not need to be running first.

This could be extended to other platforms, e.g. AppleScript on macOS.

When Ghostty develops a native API, that could be used instead to create a new window.

jcollie committed:
cli/gtk: add +new-window action

This will (on GTK) use a D-Bus method call to tell a running
Ghostty instance to open a new window. If D-Bus activation is
configured properly, Ghostty does not need to be running first.

This could be extended to other platforms, e.g. AppleScript on macOS
eventually.

When Ghostty develops a native API, that could be used instead to create
a new window.

mitchellh commented:
I'm probably going to merge #7891 first because stuff like this conflicts pretty hard with it (since I made the `+action` CLI parser generic and moved the Ghostty CLI to `src/cli/ghostty.zig`). Just a heads up that you'll have to content with some conflicts if that gets merged first. Its not messy, but Git gets confused.

jcollie committed:
cli/gtk: clarify +new-window documentation and improve instance discovery

- Add a `GHOSTTY_CLASS` environment variables to any command executed by
  Ghostty to make discovering the correct application ID easier.

- Add a flag to force the relelase application ID.

- Ensure that CLI flags to `+new-window` are mutually exclusive.

- Fix documentation about D-Bus activation requirements.

tristan957 reviewed (commented):
I don't do any custom class shenanigans, but overall the code looks good!

jcollie committed:
cli/gtk: move GTK-specific code to a new file in a subdirectory

jcollie commented on src/cli/new_window.zig:
```diff
@@ -0,0 +1,270 @@
+
+    // Use a D-Bus method call to open a new window on GTK.
+    // See: https://wiki.gnome.org/Projects/GLib/GApplication/DBusAPI
+    if (comptime build_config.app_runtime == .gtk) {
```
Done, except I moved it to `src/cli/new_window/gtk.zig`.

jcollie committed:
cli/gtk: add -e to +new-window

This adds the `-e` flag to the `+new-window` CLI action. This allows a
command to be passed from the CLI to the running instance of Ghostty.
Nothing is done with that command besides logging its presence.

jcollie committed:
cli/gtk: move actual IPC code tp apprt

mitchellh reviewed (changes requested):
This is very clean, very well done. I have some design change requests but this is heading in a great direction.

jcollie commented on src/cli/new_window.zig:
```diff
@@ -0,0 +1,186 @@
+    defer arena.deinit();
+    const alloc = arena.allocator();
+
+    if (@hasDecl(apprt.IPC, "openNewWindow")) {
```
Yep, was already working on it, but I had some chores to take care of before I finished that up.

jcollie committed:
cli/gtk: replace @hasDecl for performAction-style API

Instead of using @hasDecl, use a performAction-stype API. The C
interface for interfacing with macOS (or any other apprt where Ghostty
is embedded) is unfinished.

jcollie committed:
cli/gtk: add some more C bits

jcollie committed:
cli/gtk: remove --release and --debug flags, use optional for arguments

jcollie commented on src/cli/new_window.zig:
```diff
@@ -0,0 +1,186 @@
+    release: bool = false,
+
+    /// If `true`, open up a new window in a debug instance of Ghostty.
+    debug: bool = false,
```
Done.

jcollie committed:
cli/gtk: don't set GHOSTTY_CLASS

jcollie committed:
cli/gtk: clean ups and better error handling in GTK new-window IPC

mitchellh reviewed (changes requested):
Good progress!

jcollie committed:
cli/gtk: move IPC/sendIPC to App/performIpc

mitchellh reviewed (approved):
Awesome. This is really clean. I'd be happy to clean up anything else we find in incremental PRs. Thanks.

jcollie commented:
> This should be the default action, right? I was watching this PR because of [this discussion](https://github.com/ghostty-org/ghostty/discussions/7708#discussioncomment-13733167), but this does not solve the issue that running `ghostty` by itself from a hotkey daemon does nothing (I am still not sure why that makes a difference).

Please do not use PRs as a support forum. Open a new discussion and reference this PR.
